[takealook] step1

takealook97 opened the pull request:
controller, repository, user 패키지 생성
입력받은 정보를 repository에 저장

takealook97 committed:
initial commit

controller, repository, user 패키지 생성
입력받은 정보를 repository에 저장

dokkisan commented:
패키지 생성하셨으니 커밋 메시지를 'add: controller, repository, user 패키지 생성' 같이 수정해보시면 어떨까요..?!

dokkisan commented:
> 입력받은 정보를 repository에 저장

이 말씀이 어떤 의미인지 궁금합니다!

dokkisan commented:
mustache를 사용하면 확장자도 .html이 아닌 .mustache로 변경되나보네요! 저는 타임리프를 사용했는데, mustache 사용 후기가 궁금합니다 ㅎ ㅎ

takealook97 committed:
미션 1 구조설계

전반적인 구조설계까지밖에 못했다.
UserDTO를 통해 객체를 생성하고 이를 주고받는 구조
데이터는 DB가 아닌 메모리에 ConcurrentHashMap 으로 저장

takealook97 commented:
# 스프링 카페 1단계

안녕하세요 이번에 리뷰를 받게된 takealook 이라고 합니다.
앞으로 5주 동안 잘 부탁드립니다..!
스프링 자체를 처음 접하다보니 개념학습이 너무 부족하여 구현에 많은 어려움이 따르는 것 같습니다.
현재 전체적인 구조 설계까지 밖에 못한 상황입니다.
더 열심히해서 다음주에는 더 완성도있는 pr을 보내겠습니다.

### 진행상황
- 김영한T 스프링 기본편 ~(절반)
- 김영한T스프링 MVC1  ~1/3
- 미션 구현 step1 진행중

- 아직 까지 전반적인 로직을 이해하지 못한 상황입니다.

ksundong commented on src/main/.DS_Store:
이런 파일들은 버전관리 대상이 아닙니다.
버전관리 대상이 되지 않도록, `gitignore` 를 학습해보시고, `gitignore` 된 항목을 어떻게 `git` 에서 제거할 수 있는지 학습해보세요~

ksundong commented on src/main/java/kr/codesqaud/cafe/README.md:
empty 파일로 두신 이유가 있으신가요?

ksundong commented on src/main/java/kr/codesqaud/cafe/config/MvcConfiguration.java:
```diff
@@ -0,0 +1,7 @@
+package kr.codesqaud.cafe.config;
+
+import org.springframework.web.servlet.config.annotation.WebMvcConfigurer;
+
+public class MvcConfiguration implements WebMvcConfigurer {
+
+}
```
사용되지 않는 `Configuration` 파일은 제거해주세요.

ksundong commented on src/main/java/kr/codesqaud/cafe/controller/UserController.java:
```diff
@@ -0,0 +1,58 @@
+import org.springframework.stereotype.Controller;
+import org.springframework.ui.Model;
+import org.springframework.web.bind.annotation.*;
+
+import java.util.List;
```
```suggestion
import java.util.List;
```
`import`는 떨어지지 않도록 작성해주세요.

ksundong commented on src/main/java/kr/codesqaud/cafe/controller/UserController.java:
```diff
@@ -0,0 +1,58 @@
+public class UserController {
+    private final MemoryUserRepository memoryUserRepository;
+
+    @Autowired
```
`@Autowired`는 현재는 제한적인 경우에만 사용되는 애노테이션입니다.
`@Autowired` 를 제거하고 동작하도록 만들어보세요.

ksundong commented on src/main/java/kr/codesqaud/cafe/controller/UserController.java:
```diff
@@ -0,0 +1,58 @@
+        this.memoryUserRepository = memoryUserRepository;
+    }
+
+    @GetMapping("/user/form")
+    public String join() {
+        return "user/form";
+    }
```
`/users/user/form` 과 같은 형태로 `url` 이 매핑될텐데요.
해당 `url` 은 중복된 리소스의 표현이 발생되어, 추후 유지보수하기 어려울 수 있습니다.
제가 제안드리는 형태는 `/users/create-user-form` 과 같은 형태입니다.

ksundong replied:
또한 메서드의 명이 해당 메서드가 하는 일을 정확하게 표현하고 볼 수 없다고 생각합니다.
`join`이 적절한지 한 번 생각해보는게 어떨까요?

ksundong commented on src/main/java/kr/codesqaud/cafe/controller/UserController.java:
```diff
@@ -0,0 +1,58 @@
+    }
+
+    @GetMapping
+    public String listAllUsers(Model model){
```
```suggestion
    public String listAllUsers(Model model) {
```
요런 코드 컨벤션을 준수할 수 있도록 IntelliJ의 Reformat 기능을 사용하거나, lint 툴을 사용해보세요.

ksundong commented on src/main/java/kr/codesqaud/cafe/controller/UserController.java:
```diff
@@ -0,0 +1,58 @@
+
+    @PostMapping("/user/create")
+    public String create(UserDTO userDTO) {
+        User user = new User(userDTO.getUserId(), userDTO.getPassword(),
+                userDTO.getName(), userDTO.getEmail());
```
`UserDTO.toUserEntity()` 와 같은 식이면 생성자를 매번 작성해주지 않아도 되지 않을까요?

ksundong commented on src/main/java/kr/codesqaud/cafe/controller/UserController.java:
```diff
@@ -0,0 +1,58 @@
+        User user = new User(userDTO.getUserId(), userDTO.getPassword(),
+                userDTO.getName(), userDTO.getEmail());
+        memoryUserRepository.save(user);
+        return "redirect:/users";
```
(질문) 왜 이렇게 작성해주셨나요?

ksundong commented on src/main/java/kr/codesqaud/cafe/controller/UserController.java:
```diff
@@ -0,0 +1,58 @@
+        return "redirect:/users";
+    }
+
+    @GetMapping("/users")
```
여기도 `/users/users` 와 같은 형태가 되어 매우 혼란스러울 수 있습니다.
변경하시는걸 제안드려요.

ksundong commented on src/main/java/kr/codesqaud/cafe/controller/UserController.java:
```diff
@@ -0,0 +1,58 @@
+    }
+
+    @GetMapping("/users")
+    public String listPage(final Model model) {
```
여기는 왜 `final` 을 사용해주셨나요?

ksundong commented on src/main/java/kr/codesqaud/cafe/controller/UserController.java:
```diff
@@ -0,0 +1,58 @@
+        return "/user/list";
+    }
+
+//    @GetMapping("/users/{userId}")
+//    public String viewUserProfile(@PathVariable final String userId, final Model model) {
+//        User findUser = userService.findOne(userId).get();
+//        model.addAttribute("user", findUser);
+//        return "/user/profile";
+//    }
```
사용하지 않는 코드로 보이는데, 왜 이렇게 주석처리된 상태로 작업이 되었나요?
앞으로는 되도록이면 주석 처리된 코드는 특별한 이유가 아니면 올리지 않도록 작업하는 습관을 들여주세요.

ksundong commented on src/main/java/kr/codesqaud/cafe/dto/UserDTO.java:
```diff
@@ -0,0 +1,40 @@
+package kr.codesqaud.cafe.dto;
+
+public class UserDTO {
```
왜 `DTO` 라는 이름을 사용해주었나요? `DTO`가 가지는 의미에 대해서 알고 계신가요?

ksundong commented on src/main/java/kr/codesqaud/cafe/repository/MemoryUserRepository.java:
```diff
@@ -0,0 +1,30 @@
+import java.util.*;
+import java.util.concurrent.ConcurrentHashMap;
+
+@Repository
+public class MemoryUserRepository implements UserRepository {
+    private final Map<String, User> userRepository = new ConcurrentHashMap<>();
+    private static Long sequence = 0L;
+    @Override
+    public User save(User user) {
+        user.setSequence(++sequence);
+        userRepository.put(user.getUserId(), user);
+        return user;
+    }
+
+    @Override
+    public User findById(String id) {
+        return userRepository.get(id);
+    }
+
+    @Override
+    public List<User> findAll() {
+        return null;
+    }
+
+}
```
이 클래스에는 두 가지 문제가 있습니다. 사실 문제는 한가지가 될 수 있지만 잘 인지하지 못하고 사용하시는 것 같아서 짚어드릴게요.
1. `Repository` 가 `static` 하지 않습니다. 이럴 경우 어떤 문제가 생길 수 있을까요? 또한 작성하신 코드가 잘 동작하는 이유는 무엇일까요?
2. `ConcurrentHashMap` 을 사용해주셨지만, `Long` 을 사용해주셨는데, 이럴 경우 어떤 문제가 생길 수 있을까요?

ksundong commented on src/main/java/kr/codesqaud/cafe/repository/MemoryUserRepository.java:
```diff
@@ -0,0 +1,30 @@
+@Repository
+public class MemoryUserRepository implements UserRepository {
+    private final Map<String, User> userRepository = new ConcurrentHashMap<>();
+    private static Long sequence = 0L;
+    @Override
```
```suggestion
    private static Long sequence = 0L;

    @Override
```

ksundong commented on src/main/java/kr/codesqaud/cafe/repository/MemoryUserRepository.java:
```diff
@@ -0,0 +1,30 @@
+        return userRepository.get(id);
+    }
+
+    @Override
+    public List<User> findAll() {
+        return null;
+    }
```
이 코드는 제대로 동작하지 않을 것 같은데요. 의도한 방식이 맞나요?

ksundong commented on src/main/java/kr/codesqaud/cafe/repository/UserRepository.java:
```diff
@@ -0,0 +1,13 @@
+package kr.codesqaud.cafe.repository;
+
+import kr.codesqaud.cafe.user.User;
+
+import java.util.List;
```
```suggestion
import kr.codesqaud.cafe.user.User;
import java.util.List;
```

ksundong commented on src/main/java/kr/codesqaud/cafe/repository/UserRepository.java:
```diff
@@ -0,0 +1,13 @@
+public interface UserRepository {
+    User save(User user);
+
+    User findById(String id);
```
만약 `id`에 해당하는 `User`가 없다면 어떻게 되나요?
사용하는 쪽에서 무엇을 조심해야할까요?

ksundong commented on src/main/java/kr/codesqaud/cafe/service/UserService.java:
```diff
@@ -0,0 +1,28 @@
+import java.util.List;
+import java.util.Optional;
+
+public class UserService {
+    private final UserRepository userRepository;
+
+    public UserService(UserRepository userRepository) {
+        this.userRepository = userRepository;
+    }
+
+    public void join(User user) {
+        userRepository.save(user);
+    }
+
+    public List<User> findUsers() {
+        return new ArrayList<>(userRepository.findAll());
+    }
+
+    public User findOne(String userId) {
+        return userRepository.findById(userId);
+    }
+}
```
제대로 working 하는 코드인가요?

ksundong commented on src/main/java/kr/codesqaud/cafe/user/User.java:
```diff
@@ -0,0 +1,67 @@
+    private String password;
+    private String userName;
+    private String userEmail;
+
+
```
```suggestion

```

ksundong commented on src/main/java/kr/codesqaud/cafe/user/User.java:
```diff
@@ -0,0 +1,67 @@
+package kr.codesqaud.cafe.user;
+
+public class User {
```
해당 클래스는 동일성, 동등성 판단을 어떻게 할 수 있을까요?
한 번 생각해보세요.

ksundong commented on src/main/resources/static/user/login.html:
```diff
@@ -63,9 +63,9 @@
-                <li><a href="../user/form.html" role="button">회원가입</a></li>
+                <li class="active"><a href="../../templates/index.html">Posts</a></li>
+                <li><a href="/login.html" role="button">로그인</a></li>
+                <li><a href="ml" role="button">회원가입</a></li>
```
`ml` 이 뭔가요??

ksundong commented on src/main/resources/static/user/login_failed.html:
```diff
@@ -63,9 +63,9 @@
-                <li><a href="../user/login.html" role="button">로그인</a></li>
-                <li><a href="../user/form.html" role="button">회원가입</a></li>
+                <li class="active"><a href="../../templates/index.html">Posts</a></li>
+                <li><a href="tml" role="button">로그인</a></li>
```
`tml` 은 뭔가용???

ksundong commented on src/main/resources/templates/index.html:
```diff
@@ -0,0 +1,139 @@
+                            </strong>
+                            <div class="auth-info">
+                                <i class="icon-add-comment"></i>
+                                <span class="time">2016-01-15 18:47</span>
```
요건 나중에 채워질 예정인가요? `TODO` 를 활용하면 좋겠네요.

ksundong commented on src/main/resources/templates/index.html:
```diff
@@ -0,0 +1,139 @@
+<script src="/js/bootstrap.min.js"></script>
+<script src="/js/scripts.js"></script>
+</body>
+</html>
```
`No newline at end of file` 이 노출되는데 왜 이러는건지 한 번 학습해보시기 바랍니다.

ksundong reviewed (changes requested):
안녕하세요~ takealook! 반갑습니다. 리뷰어 디온입니다.

사실 스프링 자체는 거의 대부분이 처음 접하시니 너무 조급한 마음 가지지 않고 작은 것 부터 잘 만들어 나가시면 좋을 것 같습니다.
구조적인 설계는 전체를 만들어보고 그 때 다시 만들어보는 것도 방법입니다. 여러가지 전략이 있으니 여러가지를 모두 시도해보시기 바랍니다.

PR을 올릴 때, 코멘트로 본인이 전달하고 싶은 내용을 전달하셔도 됩니다.
이 부분이 궁금하다거나, 더 좋은 코드를 만들면 어떻게 해야할지 등등을 코드에 물어보시면 좀 더 좋은 답변을 받으실 수 있을겁니다.

또, 커밋 메시지가 지나치게 이해하기 어렵고, 너무 커밋단위가 큽니다.
[좋은 커밋 메시지에 관한 글](https://meetup.nhncloud.com/posts/106)
[좋은 커밋에 관한 글](https://github.blog/2022-06-30-write-better-commits-build-better-projects/)
[Atomic Commit](https://www.aleksandrhovhannisyan.com/blog/atomic-git-commits/)

좋은 PR description을 작성하는 방법은 다음의 글을 살펴보시기 바랍니다.
https://soojin.ro/review/cl-descriptions

또한 바로 개발에 들어가기 보다는, 미리 어떻게 할것인지 로직을 이해하는 것이 중요합니다.
이해가 가지 않는다면 빠르게 개발해보고 폐기하고 새로 개발하시면 이해하시기 쉬울거에요.
앞으로 5주간 잘 부탁드리고, 여러가지 코멘트를 남겼으니 코멘트 참고하셔서 수정부탁드립니다. 고생하셨습니다~
